docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,754 @@
+working group.
+
+- Claims
+  - Claims are statements about a particular subject. They consist of
+    name-value pairs containing the claim name, which is a string, and
+    claim value, which is arbitrary data. Example of claims could be
+    [name="version", value=1] or [name="enclave_id", value=1111].
+- Evidence
+  - This is the data about the enclave that is produced and signed by it.
+    The SGX report would be an example of evidence.
+- Endorsements
+  - This is additional data that used in the evidence verification process,
+    but is not produced by the enclave. An example of an endorsement would be
+    the quoting enclave's identity used in SGX remote attestation, because it
+    is retrieved from Intel's servers, rather than the enclave.
+- Verifier
+  - The verifier is responsible for taking in the evidence and endorsements
+    and deciding if the enclave is trustworthy.
+- Relying party
+  - The relying party is the entity interested in communicating with an
+    enclave. The enclave must attest to the relying party before the
+    relying party can trust it. The relying party can also play the role
+    of the verifier, but it does not necessarily have to.
+
+Specification
+-------------
+
+To support custom attestation formats, this document proposes adding a plugin
+model for attestation. The Open Enclave SDK will define a set of common APIs
+that each plugin must implement. Each plugin will define an UUID to distinguish
+it from other plugins.
+
+Futhermore, there will be additional attestation "plugin aware" APIs that are
+analogous to `oe_get_report` and `oe_verify_report` called `oe_get_evidence`
+and `oe_verify_evidence` respectively. These align with the ones proposed in
+the [endorsements design doc](RemoteAttestatonCollaterals.md). There will also
+be functions for registering and unregistering plugins called
+`oe_register_attestation_plugin` and `oe_unregister_attestation_plugin`. The user
+can statically link in their desired plugin and call the register plugin function.
+The attestation data can be retrieved from the "plugin aware" analogue of
+`oe_get_report` with the desired UUID. The generated data will have the UUID in
+its header. The user can call the analogue of `oe_verify_report` to verify the
+data and the Open Enclave runtime can use this UUID to determine what plugin
+verification routine to run.
+
+### Plugin API
+
+Each plugin must implement the functions below:
+
+```C
+/**
+ * Claims struct used for claims parameters for the plugin.
+ */
+struct oe_claim_t
+{
+    char* name;
+    uint8_t* value;
+    size_t value_size;
+};
+
+/**
+ * Struct that defines the structure of each plugin. Each plugin must
+ * define an UUID for its format and implement the functions in this
+ * struct. Ideally, each plugin should provide a helper function to
+ * create this struct on the behalf of the plugin users.
+ */
+struct oe_attestation_plugin_t
```

Comment:
Technical point (the only reason I'm not +1'ing the PR right now as the rest are comments that I would include with a +1):

Actually I'm just now realizing I don't understand why getting evidence and verifying evidence are in the same plugin.
An SGX plugin for getting evidence is only needed on an SGX machine (an SGX Attester).
An SGX plugin for verifying evidence is needed on every Verifier.
You can have SGX Attesters that aren't Verifiers, and you can have Verifiers that are not SGX Attesters.
By putting them in the same plugin, this increases the TCB unnecessarily.

Reply:
I think it makes sense to separate them. For example, you could have a cluster with a bunch of TEEs producing their own evidence and you want your verifier service to handle all those types. In that case, the verifier would only care about the the `verify_evidence` part and the TEEs would only care about the `get_evidence` part.

We could have two plugins--one for getting evidence and one for verifier evidence. The UUID can be used to connect them.